### PR TITLE
Feature/architecture

### PR DESCRIPTION
**Pull request by leoTiez**
The architecture is implemented according to the paper. However, I dropped the keras implementation and used 'pure' tensorflow instead, since it is more comprehensible what actually happens. Training works and the network can achieve around 85% accuracy. Note amazing, but the hyperparameters are not yet set properly. 

Moreover, I want to investigate whether it is necessary to implement the backpropagation through time algorithm, or whether the 'standard' backprop implementation is sufficient. However, I think this is a new feature and it might be more appropriate to put it in a separate branch

### CODE REVIEW

**Comment by adrianc-a** [comment]
I'd prefer backprop over time on it's own branch

File: CRNN.py [hunk 1]
```diff
@@ -193,35 +194,87 @@ def compute_output_shape(self, input_shape):
...
     model.add(tf.keras.layers.Flatten())
-    model.add(tf.keras.layers.Dense(50, activation='relu'))
     model.add(tf.keras.layers.Dense(10, activation='softmax'))
+
     model.summary()
 
     model.compile(
         optimizer=tf.train.AdagradOptimizer(learning_rate=.1),
-        loss='sparse_categorical_crossentropy',
+        loss='kullback_leibler_divergence',
```
Comment (adrianc-a) [inline CRNN.py]: why have we changed this? the orig. paper uses cross-entropy.

File: RCNN_tf.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+    if padding.upper() not in PADDING_LIST:
+        raise ValueError('Padding value is not understood')
+
+    output = tf.nn.conv2d(input_data, weights, strides=strides, padding=padding.upper())
+
+    return output
+
+
+def pooling_layer(
```
Comment (adrianc-a) [inline RCNN_tf.py:101]: I get that you want to use 'pure' tf, but i feel like this introduces more potential for error. what is wrong with using [tf.nn.pool](https://www.tensorflow.org/api_docs/python/tf/nn/pool) ?

File: RCNN_tf.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    return output
+
+
+def global_max_pooling_layer(
+        input_data
+):
+    return tf.reduce_max(input_data, axis=[1, 2])
+
+
+def softmax_layer(
```
Comment (adrianc-a) [inline RCNN_tf.py:123]: again, what is the advantage/reason over `tf.nn.softmax`?

File: RCNN_tf.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+        tf.truncated_normal([num_output_dim], stddev=std),
+        trainable=True,
+        name=name + '_bias'
+    )
+
+    linear_trans = tf.matmul(input_data, weights) + bias
+    return tf.nn.softmax(linear_trans), linear_trans
+
+
+def accuracy(
```
Comment (adrianc-a) [inline RCNN_tf.py:146]: `tf.metrics.accuracy` ?

File: RCNN_tf.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+        tf.argmax(labels, axis=1),
+        tf.argmax(result, axis=1)
+    )
+
+    return tf.reduce_mean(tf.cast(correct_prediction, PRECISION_TF))
+
+
+if __name__ == '__main__':
+    # Setting the parameters
+    input_shape_ = [None, 28, 28, 1]
```
Comment (adrianc-a) [inline RCNN_tf.py:155]: realistically, all the problem specific dimensions need to be abstracted for now it's probably ok, but i guess keep it in mind.

File: RCNN_tf.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+    output_shape_ = [None, 10]
+    learning_rate_ = .1
+    epochs_ = 5
```
Comment (adrianc-a) [inline RCNN_tf.py:158]: also more a comment for the future, we should just save checkpoints and make this not so fixed.

File: RCNN_tf.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+    # Flatten tensor for softmax layer
+    flatten_ = tf.reshape(global_max_, [-1, global_max_.shape[-1].value])
+
+    # Softmax layer
+    result_, linear_trans_ = softmax_layer(
+        flatten_,
+        global_max_.shape[-1].value,
+        output_shape_[1]
+    )
+
```
Comment (adrianc-a) [inline RCNN_tf.py:299]: afaik, the architecture looks as it does in the paper, good.

File: RCNN_tf.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+    # Max pooling layer
+    global_max_ = global_max_pooling_layer(rcl_layer_4_)
+
+    # Flatten tensor for softmax layer
+    flatten_ = tf.reshape(global_max_, [-1, global_max_.shape[-1].value])
+
+    # Softmax layer
+    result_, linear_trans_ = softmax_layer(
```
Comment (adrianc-a) [inline RCNN_tf.py:294]: clever

File: RCNN_tf.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+                                      input_placeholder_: test_data_,
+                                      output_placeholder_: test_labels_,
+                                      rate_placeholder_: 0
+                                  })
+            print "\nEpoch:", (epoch + 1), \
+                "cost =", "{:.3f}".format(avg_cost_),\
+                "test accuracy: {:.3f}".format(test_acc_)
+
+        print "\nTraining complete!"
+        print sess_.run(accuracy_,
```
Comment (adrianc-a) [inline RCNN_tf.py:355]: this can really be simplified using `tf.data.Dataset` which can handle shuffling, batching, repeating etc.

File: load_data.py [hunk 10]
```diff
@@ -34,19 +36,19 @@ def load_cifar(file_name, dtype, file_path='./data/cifar-10-batches-py/'):
...
     std = np.std(cifar_data, axis=0)
     batch_size = cifar_data.shape[0]
 
-    cifar_data = (cifar_data - mean) / (std - np.finfo(dtype).eps)
+    # cifar_data = (cifar_data - mean) / (std - np.finfo(dtype).eps)
 
-    R = cifar_data[:, 0:1024]
-    G = cifar_data[:, 1024:2048]
-    B = cifar_data[:, 2048:]
+    R = cifar_data[:, 0:1024] / 255.
```
Comment (adrianc-a) [inline load_data.py:53]: not super pertinent to this pr, but i feel like we should standardize image preprocessing

**Review by adrianc-a — CHANGES_REQUESTED** [review]

**Comment by adrianc-a** [comment]
Also, this whole CRNN/RCNN naming scheme with two different main's is i think convoluted. since this is it's own branch and you're trying to fix CRNN.py, why don't we just overwrite that?

File: CRNN.py (hunk 1 above)
Comment (leoTiez) [reply CRNN.py]: Was a test. I will change it. However, we should use the `RCNN_tf` implementation anyways.

File: RCNN_tf.py (hunk 2 above)
Comment (leoTiez) [reply RCNN_tf.py:101]: Take a look some lines below :). This is just to provide a simple interface

File: RCNN_tf.py (hunk 3 above)
Comment (leoTiez) [reply RCNN_tf.py:123]: Again, `tf.nn.softmax` is used. But we need to compute the linear transformation first.

File: RCNN_tf.py (hunk 5 above)
Comment (leoTiez) [reply RCNN_tf.py:155]: I don't understand what you mean. I what way abstracting it?

File: load_data.py (hunk 10 above)
Comment (leoTiez) [reply load_data.py:53]: Depends on what you mean with standardizing. All Images of all data sets should be preprocessed the same way, I totally agree. We only need to do this consistently. Which normalization we use is less important I find.

File: RCNN_tf.py (hunk 9 above)
Comment (leoTiez) [reply RCNN_tf.py:355]: I will take a look on that

File: RCNN_tf.py (hunk 4 above)
Comment (leoTiez) [reply RCNN_tf.py:146]: I will take a look on that.

**Comment by leoTiez** [comment]
I can remove the old implementation. However, I thought it would be nice to provide a Keras implementation to the community.

File: RCNN_tf.py (hunk 5 above)
Comment (adrianc-a) [reply RCNN_tf.py:155]: we should have separate main functions, and the network would have to have different input dimension, but i think we can save this for later.

File: load_data.py (hunk 10 above)
Comment (adrianc-a) [reply load_data.py:53]: yes, just that it should be the same across the datasets.

File: RCNN_tf.py (hunk 5 above)
Comment (leoTiez) [reply RCNN_tf.py:155]: :+1:

File: RCNN_tf.py [hunk 11]
```diff
@@ -169,28 +164,44 @@ def accuracy(
...
 
     # Create input and output placeholder
     input_placeholder_ = tf.placeholder(PRECISION_TF, input_shape_)
     output_placeholder_ = tf.placeholder(PRECISION_TF, output_shape_)
+
+    # Create data set objects
+    training_data_set_ = tf.data.Dataset.from_tensor_slices((
+        input_placeholder_,
+        output_placeholder_
+    )).repeat().batch(batch_size=batch_size_)
```
Comment (adrianc-a) [inline RCNN_tf.py]: i think we should also shuffle it.

**Review by adrianc-a — APPROVED** [review]
let's tidy it up later.
